sklearn/inspection/_ice.py

```diff
@@ -0,0 +1,832 @@
+"""Individual Conditional Expectation (ICE) plots for regression and
+classification models."""
+
+# Authors: Peter Prettenhofer
+#          Trevor Stephens
+#          Nicolas Hug
+#          Madhura Jayaratne
+# License: BSD 3 clause
+
+from collections.abc import Iterable
+from itertools import count
+from itertools import chain
+import numbers
+import warnings
+
+import numpy as np
+from scipy import sparse
+from scipy.stats.mstats import mquantiles
+from joblib import Parallel, delayed
+
+from sklearn.ensemble._hist_gradient_boosting.gradient_boosting import (
+    BaseHistGradientBoosting)
+from ..base import is_classifier
+from ..base import is_regressor
+from ..ensemble._gb import BaseGradientBoosting
+from ..exceptions import NotFittedError
+from ..pipeline import Pipeline
+from ..utils import _determine_key_type
+from ..utils import _get_column_indices
+from ..utils import _safe_indexing
+from ..utils import check_array
+from ..utils import check_matplotlib_support  # noqa
+from ..utils.extmath import cartesian
+from ..utils.validation import check_is_fitted
+
+
+__all__ = ['individual_conditional_expectation',
+           'plot_individual_conditional_expectation',
+           'IndividualConditionalExpectationDisplay']
+
+
+def _grid_from_X(X, percentiles, grid_resolution):
```

Comment:
I cannot review right now but it seems really redundant. PDP and ICE should share code and we should factorize code as much as possible. you can create a `_base.py` file to import common part.

Reply:
Thanks, @glemaitre. I have extracted a number of common code into methods inside `_ice.py` which are being used by both PDP (https://github.com/madhuracj/scikit-learn/blob/06b1c617f9a28d59822d4346d1087d652ef9631d/sklearn/inspection/_partial_dependence.py#L16-L18) and ICE. 

For example, see https://github.com/madhuracj/scikit-learn/blob/06b1c617f9a28d59822d4346d1087d652ef9631d/sklearn/inspection/_partial_dependence.py#L188-L191 and https://github.com/madhuracj/scikit-learn/blob/06b1c617f9a28d59822d4346d1087d652ef9631d/sklearn/inspection/_ice.py#L353-L356.

I agree it makes more sense to have them in `_base.py` rather than `_ice.py`. Will update the PR with that.

Reply:
Done in https://github.com/scikit-learn/scikit-learn/pull/16164/commits/2da2d4ccc1e668d44f589a4e45eae1440e91b474